File: producer_consumer_semaphore.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <semaphore.h>
#include <string.h>
#include <unistd.h>

#define BUFFER_SIZE 5
#define MAX_STRING_LEN 10 // Maximum length for the produced string

char buffer[BUFFER_SIZE][MAX_STRING_LEN]; // Buffer to hold strings
int in = 0, out = 0;
sem_t bfull, bempty; // Semaphore for full and empty slots

// Function to produce a string
void* producer(void* arg) {
    char item[MAX_STRING_LEN];
    for (int i = 0; i < 10; i++) { // Produce 10 strings
        sem_wait(&bempty); // Wait for empty slot
        snprintf(item, MAX_STRING_LEN, "Item%d", i); // Produce an item within the max length
        strncpy(buffer[in], item, MAX_STRING_LEN); // Add string to buffer
        printf("Produced: %s\n", item);
        in = (in + 1) % BUFFER_SIZE; // Circular increment
        sem_post(&bfull); // Signal that there is a new item
        sleep(rand() % 2); // Sleep for a random time
    }
    return NULL;
}

// Function to consume a string
void* consumer(void* arg) {
    for (int i = 0; i < 10; i++) { // Consume 10 strings
        sem_wait(&bfull); // Wait for a full slot
        char item[MAX_STRING_LEN];
        strncpy(item, buffer[out], MAX_STRING_LEN); // Get string from buffer
        printf("Consumed: %s\n", item);
        out = (out + 1) % BUFFER_SIZE; // Circular increment
        sem_post(&bempty); // Signal that there is an empty slot
        sleep(rand() % 2); // Sleep for a random time
    }
    return NULL;
}

int main() {
    pthread_t prod, cons;
    sem_init(&bfull, 0, 0); // Initialize semaphore bfull to 0
    sem_init(&bempty, 0, BUFFER_SIZE); // Initialize semaphore bempty to BUFFER_SIZE

    // Create producer and consumer threads
    pthread_create(&prod, NULL, producer, NULL);
    pthread_create(&cons, NULL, consumer, NULL);

    // Join threads
    pthread_join(prod, NULL);
    pthread_join(cons, NULL);

    sem_destroy(&bfull); // Destroy semaphore
    sem_destroy(&bempty); // Destroy semaphore
    return 0;
}

File: producer_consumer_semaphore.py
import threading
import random
import time

BUFFER_SIZE = 5
ITEMS_TO_PRODUCE = 10
PRODUCER_COUNT = 2
CONSUMER_COUNT = 2

# Shared buffer and synchronization primitives
buffer = [None] * BUFFER_SIZE
in_index = 0  # Points to the next insertion slot
out_index = 0  # Points to the next removal slot

# Semaphores
empty_slots = threading.Semaphore(BUFFER_SIZE)  # Tracks available slots
full_slots = threading.Semaphore(0)  # Tracks filled slots
buffer_lock = threading.Lock()  # Protects buffer access


# Producer function
def producer(producer_id):
    global in_index
    for _ in range(ITEMS_TO_PRODUCE):
        item = random.randint(1, 100)  # Produce a random number

        empty_slots.acquire()  # Wait for an empty slot
        with buffer_lock:  # Ensure exclusive access to the buffer
            buffer[in_index] = item
            print(f"Producer {producer_id} inserting {item} in slot {in_index}")
            in_index = (in_index + 1) % BUFFER_SIZE

        full_slots.release()  # Signal that a new item is available
        time.sleep(random.uniform(0.5, 1.5))  # Simulate work


# Consumer function
def consumer(consumer_id):
    global out_index
    for _ in range(ITEMS_TO_PRODUCE):
        full_slots.acquire()  # Wait for a filled slot
        with buffer_lock:  # Ensure exclusive access to the buffer
            item = buffer[out_index]
            print(f"Consumer {consumer_id} consuming {item} from slot {out_index}")
            out_index = (out_index + 1) % BUFFER_SIZE

        empty_slots.release()  # Signal that a slot is now available
        time.sleep(random.uniform(0.5, 1.5))  # Simulate work


# Main function
def main():
    producers = []
    consumers = []

    # Create producer threads
    for i in range(PRODUCER_COUNT):
        thread = threading.Thread(target=producer, args=(i + 1,))
        producers.append(thread)
        thread.start()

    # Create consumer threads
    for i in range(CONSUMER_COUNT):
        thread = threading.Thread(target=consumer, args=(i + 1,))
        consumers.append(thread)
        thread.start()

    # Wait for all threads to complete
    for thread in producers:
        thread.join()

    for thread in consumers:
        thread.join()


if __name__ == "__main__":
    main()
